examples/structured_data/tabtransformer.py

```diff
@@ -0,0 +1,537 @@
+    return model
+
+
+baseline_model = create_baseline_model(
+    embedding_dims, num_mlp_blocks, mlp_hidden_units_factors, dropout_rate
+)
+
+print("Total model weights:", baseline_model.count_params())
+keras.utils.plot_model(baseline_model, show_shapes=True, rankdir="LR")
+
+"""
+Let's train and evaluate the baseline model:
+"""
+
+history = run_experiment(baseline_model)
+
+"""
+The baseline linear model achieves ~81% validation accuracy.
+"""
+
+"""
+## Experiment 2: TabTransformer
+
+The TabTransformer architecture works as follows:
+
+1. All the categorical features are encoded as embeddings, using the same `embedding_dims`.
+This means that each value in each categorical feature will have its own embedding vector.
+2. A column embedding, one embedding vector for each categorical feature, is added (point-wise) to the categorical feature embedding.
+3. The embedded categorical features are fed into a stack of Transformer blocks.
+Each Transformer block consists of a multi-head self-attention layer followed by a feed-forward layer.
+3. The outputs of the final Transformer layer, which are the *contextual embeddings* of the categorical features,
+are concatenated with the input numerical features, and fed into a final MLP block.
+4. A `softmax` classifer is applied at the end of the model.
+
+The [paper](https://arxiv.org/abs/2012.06678) discusses both addition and concatenation of the column embedding in the
+*Appendix: Experiment and Model Details* section.
+The architecture of TabTransformer is shown below, as presented in the paper.
+
+![tabtransformer](https://github.com/keras-team/keras-io/blob/master/examples/structured_data/img/tabtransfoermer/tabtransfoermer.png)
+"""
+
+
+def create_tabtransformer_classifier(
+    num_transformer_blocks,
+    num_heads,
+    embedding_dims,
+    mlp_hidden_units_factors,
+    dropout_rate,
+    use_column_embedding=False,
+):
+
+    # Create model inputs.
+    inputs = create_model_inputs()
+    # encode features.
+    encoded_categorical_feature_list, numerical_feature_list = encode_inputs(
+        inputs, embedding_dims
+    )
+    # Stack categorical feature embeddings for the Tansformer.
+    encoded_categorical_features = tf.stack(encoded_categorical_feature_list, axis=1)
+    # Concatenate numerical features.
+    numerical_features = layers.concatenate(numerical_feature_list)
+
+    # Add column embedding to categorical feature embeddings.
+    if use_column_embedding:
+        num_columns = encoded_categorical_features.shape[1]
+        column_embedding = layers.Embedding(
+            input_dim=num_columns, output_dim=embedding_dims
+        )
+        column_indices = tf.range(start=0, limit=num_columns, delta=1)
+        encoded_categorical_features = encoded_categorical_features + column_embedding(
+            column_indices
+        )
+
+    # Create multiple layers of the Transformer block.
+    for block in range(num_transformer_blocks):
```

Comment:
Likewise it would be useful to have an indication that `block` is a counter (e.g. `block_index` or something) rather than an actual block/layer object

Reply:
Done